spec/structured-data.html

```diff
@@ -107,5 +107,50 @@ <h1><ins>Runtime Semantics: BuildImmutableProperty ( _value_, _name_, _reviver_
         </ul>
       </emu-note>
     </emu-clause>
+
+    <emu-clause id="sec-json.stringify">
+      <h1>JSON.stringify ( _value_ [ , _replacer_ [ , _space_ ] ] )</h1>
+
+      <emu-clause id="sec-serializejsonproperty" aoid="SerializeJSONProperty">
+        <h1>Runtime Semantics: SerializeJSONProperty ( _state_, _key_, _holder_ )</h1>
+        <p>The abstract operation SerializeJSONProperty takes arguments _state_, _key_, and _holder_. It performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _value_ be ? Get(_holder_, _key_).
+          1. If Type(_value_) is Object or BigInt, then
+            1. Let _toJSON_ be ? GetV(_value_, *"toJSON"*).
+            1. If IsCallable(_toJSON_) is *true*, then
+              1. Set _value_ to ? Call(_toJSON_, _value_, &laquo; _key_ &raquo;).
+          1. If _state_.[[ReplacerFunction]] is not *undefined*, then
+            1. Set _value_ to ? Call(_state_.[[ReplacerFunction]], _holder_, &laquo; _key_, _value_ &raquo;).
+          1. If Type(_value_) is Object, then
+            1. If _value_ has a [[NumberData]] internal slot, then
+              1. Set _value_ to ? ToNumber(_value_).
+            1. Else if _value_ has a [[StringData]] internal slot, then
+              1. Set _value_ to ? ToString(_value_).
+            1. Else if _value_ has a [[BooleanData]] internal slot, then
+              1. Set _value_ to _value_.[[BooleanData]].
+            1. Else if _value_ has a [[BigIntData]] internal slot, then
+              1. Set _value_ to _value_.[[BigIntData]].
+          1. <ins>If Type(_value_) is Record or Type(_value_) is Tuple, set _value_ to ! ToObject(_value_).</ins>
```

Comment:
i'm confused why this approach - creating an unnecessary object - instead of:
```suggestion
          1. <ins>If ! IsRecord(_value_) is *true*, set _value_ to ! thisRecordValue(_value_).</ins>
          1. <ins>If ! IsTuple(_value_) is *true*, set _value_ to ! thisTupleValue(_value_).</ins>
```

Reply:
The reason to wrap rather than unwrap records and tuples is that all the abstract operations used by `JSON.stringify` (such as `LengthOfArrayLike` and `EnumerableOwnProperties`) work with objects and not with primitive values.

I initially tried to avoid wrapping them (https://github.com/tc39/proposal-record-tuple/commit/b1b503892cb4cd19e7915783f8fd6d05d4d773c0), but then @littledan pointed out (https://github.com/tc39/proposal-record-tuple/pull/163#discussion_r455890071) that I would need to wrap them anyway in order to use `[[OwnPropertyNames]]()` on records.

Reply:
It seems like a cleaner fix to me to change the internals of the spec to support R&T where they support Objects, as appropriate.

Reply:
As @nicolo-ribaudo has indicated elsewhere, those sorts of changes may have broader observable effects, so it'd be good to pursue them separately.

IMO the wording here is nice and clear, but the most important thing is that it "works". We can land this patch and iterate on further editorial changes if people have other ideas, all the way through Stage 4 and beyond.